lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp

```diff
@@ -1120,42 +1167,38 @@ struct AIEObjectFifoStatefulTransformPass
     DMAChannelAnalysis dmaAnalysis(device);
     OpBuilder builder = OpBuilder::atBlockEnd(device.getBody());
     auto ctx = device->getContext();
-
-    //===------------------------------------------------------------------===//
-    // Create objectFifos
-    //===------------------------------------------------------------------===//
     std::set<TileOp>
         objectFifoTiles; // track cores to check for loops during unrolling
 
-    for (auto createOp : device.getOps<ObjectFifoCreateOp>()) {
-      objectFifoTiles.insert(createOp.getProducerTileOp());
-      bool shared = false;
+    //===------------------------------------------------------------------===//
+    // Split objectFifos into a consumer end and producer end if needed
+    //===------------------------------------------------------------------===//
+    // We are going to create additional createObjectFifoOps, so get a copy of
+    // all "original" ones before the loop to avoid looping over newly created
+    // ones.
```

Comment:
I apologize if I'm misunderstanding again... I'm having a bit of trouble understanding the two for loops over createOFOps.

Since we create new objectFifos, we do want to loop over them again right? To ensure we create their elements too.

The way I'm reading this rn is: 
- loop once over all ops, find split opportunities and create new OFs,
- loop again over all ops and create their elements (so I don't see the use of createFifoOps).

Couldn't you instead:
- only loop once, any newly create objectFifos will be caught as well,
- for each OF, check if it requiresDMA (either it does because it has data transformation info or it is a broadcast, or it does not require a DMA because it has shared memory and no stream attributes + the bool 'shared' is equal to the result of requiresDMA),
- based on what happened above, create new objectFifos and do all the replacing needed in other ops,
- after you are done iterating over consumers, you can create the objectFifo elements for this objectFifo.

Reply:
Yes, you capture well what I'm doing currently. For me, the previous implementation took a while to understand because it changed the set of createFifoOps _as we were iterting over that same set_. I thought it would be easier to conceptually split what we are doing into two phases, like you describe, instead of trying to do everything at once:

1. Split the fifos into consumer/producer side as needed.
2. For the new set of fifos (split ones), create all their elements.

The main reason I changed it wasn't for ease of understanding though; it was to avoid infinite loops. This explanation may be a bit wordy so bear with me.

In the previous implementation, we split the objectFifo into two whenever `shared=false`, creating a new objectFifo on the consumer tile (I'm gonna call this the consumer-side objectFifo). One of the next iterations of the loop then picked up the newly created objectFifo, so we could be sure to create the buffers and locks for both ends.**

The reason you did not end up in an infinite loop is, because eventually, when the iteration got to the consumer objectFifo, this would _not_ be split again. This is because the created consumer end of the objectFifo lists the same tile as both producer and consumer, so it is considered memory-adjacent (`shared=true`). To me, it definitely took a while to see why we are not looping infinitely, repeatedly splitting up the consumer objectFifo, until I discovered this.

Now, I had to make a change that allowed `shared=false`, even if tiles were memory-adjacent. That is why I adjusted that check -- we only use shared memory if the data layout should not be transformed.

But now, in this case, `shared` will always be `false`, even for the newly created consumer side, thus leading to that side being repeatedly split up.

So, that is why I opted for the two loops. Avoid infinitely looping, but also ease of understanding. To me, this is clearer, but obviously I am very biased because I wrote it, so of course I would understand it. My reasoning to why it is clearer would be that we are not changing the set we are iterating over. At least for me, this is easier to wrap my head around (things like DFS/BFS where you change the set as you are iterating take more brain processing for me, and why add that burden when we can also express it without that?).

 **_(Side note: I am not fully convinced that the previous algorithm was provably correct; Consider the case if the new objectFifo was inserted somewhere before the current iterator, i.e. we already iterated past it. Then we would not pick it up again. Maybe you can point me to documentation about the iteration/insertion order -- something that guarantees newly Ops will be added to the end of the iterator -- but that's not the main point I'm trying to make.)_

Reply:
Thank you so much for taking the time to explain this!!

I apologize that I lacked key comments in there that would have made your understanding easier. But like you say, since I wrote it l, it was much easier for me to understand it.

And while I understand how you changed requiresDMA + isSharedMemory, I didn't realize it would create infinite loops. The two loops now definitely make sense.

By the way, I also think it is clearer to do things in steps and have a dedicated loop for each part of the lowering. I just wanted to see what your additional reasoning was and I'm glad I asked since I now understand!

I know what you mean about the objectFifo ends from the POV of this pass implementation, but users have taken advantage of using the objectFifo with same tile as both producer and consumer and it is still a valid objectFifo with its own state to track, so I think it makes sense to treat it the same.